idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/editor/folding/FoldingSyntax.kt

```diff
@@ -1,26 +1,89 @@
 package arrow.meta.ide.dsl.editor.folding
 
 import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.ide.dsl.utils.typeReferences
+import arrow.meta.ide.phases.editor.extension.ExtensionProvider
 import arrow.meta.phases.ExtensionPhase
 import com.intellij.lang.ASTNode
 import com.intellij.lang.Language
 import com.intellij.lang.folding.FoldingBuilder
 import com.intellij.lang.folding.FoldingBuilderEx
 import com.intellij.lang.folding.FoldingDescriptor
-import com.intellij.lang.folding.LanguageFolding
 import com.intellij.openapi.editor.Document
 import com.intellij.psi.PsiElement
-import com.intellij.util.KeyedLazyInstance
 import org.jetbrains.kotlin.idea.KotlinLanguage
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 interface FoldingSyntax {
+
+  /**
+   * Utility to add a [FoldingBuilder]
+   * @param isTypeMatching: fun to match on the parent type only so folding regions are identified
+   * @param toFoldString: used to create the hint for the matching folding region
```

Comment:
```suggestion
   * @param toFoldString: the resulting hint for the matched folding region
```

Reply:
For simplifying this you may rename them to
Hint: (KtTypeReference) -> String
and
match: (KtTypeReference) -> Boolean